ext/openssl/ossl_x509cert.c

```diff
@@ -704,6 +704,84 @@ ossl_x509_eq(VALUE self, VALUE other)
     return !X509_cmp(a, b) ? Qtrue : Qfalse;
 }
 
+static VALUE
+load_chained_certificates(VALUE _io) {
+    BIO *in = (BIO*)_io;
+    VALUE certificates = rb_ary_new();
+
+    X509 *x509 = NULL;
+
+    // Check the certificate format is PEM or DER:
+
+    // Case 1: certificate format is PEM:
+    if ((x509 = PEM_read_bio_X509(in, NULL, NULL, NULL)) != NULL) {
+        do {
+            rb_ary_push(certificates, ossl_x509_new(x509));
```

Comment:
`x509` leaks if `ossl_x509_new()` or `rb_ary_push()` raises an error.

Reply:
Once `ossl_x509_new` is called, it should not leak, since it will become garbage, right?

Regarding if `ossl_x509_new` fails, then yes, it would be. Is this possible? If so, we should create safe variant IMHO, i.e. `ossl_x509_new_safe`.

Reply:
`ossl_x509_new()` creates a new copy of `X509` and an OpenSSL::X509::Certificate wrapping it, so an error from `rb_ary_push()` will still cause a leak. Although both should be able to fail only due to memory allocation failure, we've been catching all cases.

Reply:
Okay, I see the issue. I moved everything inside `rb_ensure`.